File: python/semiquadratic_cost.py
"""
BSD 3-Clause License

Copyright (c) 2019, HJ Reachability Group
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Author(s): David Fridovich-Keil ( dfk@eecs.berkeley.edu )
"""
################################################################################
#
# Semiquadratic cost, derived from Cost base class. Implements a
# cost function that is flat below a threshold and quadratic above, in the
# given dimension.
#
################################################################################

import torch

from cost import Cost

class SemiquadraticCost(Cost):
    def __init__(self, dimension, threshold, oriented_right, name=""):
        """
        Initialize with dimension to add cost to and threshold above which
        to impose quadratic cost.

        :param dimension: dimension to add cost
        :type dimension: uint
        :param threshold: value above which to impose quadratic cost
        :type threshold: float
        :param oriented_right: Boolean flag determining which side of threshold
          to penalize
        :type oriented_right: bool
        """
        self._dimension = dimension
        self._threshold = threshold
        self._oriented_right = oriented_right
        super(SemiquadraticCost, self).__init__(name)

    def __call__(self, xu, k=0):
        """
        Evaluate this cost function on the given input and itme, which might
        either be a state `x` or a control `u`. Hence the input is named `xu`.
        NOTE: `xu` should be a PyTorch tensor with `requires_grad` set `True`.
        NOTE: `xu` should be a column vector.

        :param xu: state of the system
        :type xu: torch.Tensor
        :return: scalar value of cost
        :rtype: torch.Tensor
        """
        if self._oriented_right:
            if xu[self._dimension, 0] > self._threshold:
                return (xu[self._dimension, 0] - self._threshold) ** 2
        else:
            if xu[self._dimension, 0] < self._threshold:
                return (xu[self._dimension, 0] - self._threshold) ** 2

        return torch.zeros(1, 1, requires_grad=True).double()

File: src/semiquadratic_cost.cpp
/*
 * Copyright (c) 2019, The Regents of the University of California (Regents).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 *    3. Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Please contact the author(s) of this library if you have any questions.
 * Authors: David Fridovich-Keil   ( dfk@eecs.berkeley.edu )
 */

///////////////////////////////////////////////////////////////////////////////
//
// Quadratic cost in a particular (or all) dimension(s).
//
///////////////////////////////////////////////////////////////////////////////

#include <ilqgames/cost/semiquadratic_cost.h>
#include <ilqgames/utils/types.h>

#include <glog/logging.h>

namespace ilqgames {

// Evaluate this cost at the current input.
float SemiquadraticCost::Evaluate(const VectorXf& input) const {
  CHECK_LT(dimension_, input.size());

  const float diff = input(dimension_) - threshold_;
  if ((diff > 0.0 && oriented_right_) || (diff < 0.0 && !oriented_right_))
    return 0.5 * weight_ * diff * diff;

  return 0.0;
}

// Quadraticize this cost at the given input, and add to the running
// sum of gradients and Hessians (if non-null).
void SemiquadraticCost::Quadraticize(const VectorXf& input, MatrixXf* hess,
                                     VectorXf* grad) const {
  CHECK_LT(dimension_, input.size());

  // Handle no cost case first.
  const float diff = input(dimension_) - threshold_;
  if ((diff < 0.0 && oriented_right_) || (diff > 0.0 && !oriented_right_))
    return;

  // Check dimensions.
  CHECK_NOTNULL(hess);
  CHECK_NOTNULL(grad);
  CHECK_EQ(input.size(), hess->rows());
  CHECK_EQ(input.size(), hess->cols());
  CHECK_EQ(input.size(), grad->size());

  // Compute gradient and Hessian.
  float dx = weight_ * diff;
  float ddx = weight_;

  ModifyDerivatives(input, &dx, &ddx);

  (*grad)(dimension_) += dx;
  (*hess)(dimension_, dimension_) += ddx;
}

}  // namespace ilqgames
